### DISCUSSION TITLE
Is support for 'volatile' planned?  -- solved: it's now available

### DISCUSSION THREAD

**Discussion by pierre-rouleau** [vlang/v#9435] [General]
Is support for writing low-level, hardware-accessing  device driver type code planned for V where the use of a volatile keyword could be used to prevent optimization removal of multiple reads of a hardware register or a memory-mapped device?

Or is V counting on this type of code to be written in pure C?

**Comment by alichraghi** [comment]


**Reply by pierre-rouleau** [reply]
Thanks.  Perhaps but then the Q&A information will not be visible for other people to see. Is there other another place short of the V source code where I can get this information?

**Reply by pierre-rouleau** [reply]
Thanks, I just read it.  Looks to me that V is not intended to be a prime language for low-level code and therefore you'd revert to C to control access to hardware registers.  It's Ok to me, I just want to know. Correct me if I got the wrong impression.

**Reply by pierre-rouleau** [reply]
Thanks!

**Reply by pierre-rouleau** [reply]
OTOH,  the [Functions 2  - Pure functions by default section of the the manual](https://github.com/vlang/v/blob/master/doc/docs.md#pure-functions-by-default) describes the --enable-global flag stating:  *"but this is intended for low-level applications like kernels and drivers"*.   Are you planning V to be used for the development of device drivers directly or with the use of C?

**Comment by medvednikov** [comment]
The goal is to cover all domains, including kernels and drivers, so yes it'll be possible. Probably via an attribute.

**Reply by pierre-rouleau** [reply]
That makes sense. Thanks.

**Reply by pierre-rouleau** [reply]
Really this is very encouraging.

**Comment by dumblob** [comment]
@pierre-rouleau see https://github.com/vlang/v/issues/2761#issuecomment-621329529

**Reply by pierre-rouleau** [reply]
Great news!

**Comment by dumblob** [comment]
`volatile` became part of *V* - see https://github.com/vlang/v/pull/11940 .

Feel free to close this discussion (mark as "solved").

**Comment by pierre-rouleau** [comment]
Closing ( it's solved but I don't have access to adding a label; I updated the title instead)
